go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,551 @@
+			OnNewLeader:      HAClusterNewLeader,
+		},
+	}
+
+	//go leaderelection.RunOrDie(context.Background(), lec)
+	hacluster.leaderElector, err = leaderelection.NewLeaderElector(lec)
+	if err != nil {
+		return err
+	}
+
+	go hacluster.leaderElector.Run(context.Background())
+
+	return hacluster.WatchOvnDbEndpoints()
+}
+
+// ConfigureAsActive configures the node as active.
+func (hacluster *OvnHAMasterController) ConfigureAsActive(masterNodeName string) error {
+	if hacluster.manageDBServers {
+		// Step 1: Update the ovnkube-db endpoints with invalid Ip.
+		// Step 2: Promote OVN DB servers to become active
+		// Step 3: Make sure that ovn-northd has done one round of
+		//         flow computation.
+		// Step 4: Update the ovnkube-db endpoints with the new master Ip.
+
+		// Find the endpoint for the service
+		ep, err := hacluster.cluster.Kube.GetEndpoint(config.Kubernetes.OVNConfigNamespace, ovnkubeDbEp)
+		if err != nil {
+			ep = nil
+		}
+		err = hacluster.updateOvnDbEndpoints(ep, true)
+		if err != nil {
+			logrus.Errorf("%s Endpount create/update failed", ovnkubeDbEp)
+			return err
+		}
+
+		// Promote the OVN DB servers
+		err = hacluster.PromoteOVNDbs(ovnNorthDbPort, ovnSouthDbPort)
+		if err != nil {
+			logrus.Errorf("Promoting OVN ovsdb-servers to active failed")
+			return err
+		}
+
+		// Wait for ovn-northd sync up
+		err = hacluster.syncOvnNorthd()
+		if err != nil {
+			logrus.Errorf("Syncing ovn-northd failed")
+			return err
+		}
+
+		ep, err = hacluster.cluster.Kube.GetEndpoint(config.Kubernetes.OVNConfigNamespace, ovnkubeDbEp)
+		if err != nil {
+			// This should not happen.
+			ep = nil
+		}
+		err = hacluster.updateOvnDbEndpoints(ep, false)
+		if err != nil {
+			logrus.Errorf("%s Endpount create/update failed", ovnkubeDbEp)
+			return err
+		}
+	}
+
+	// run the cluster controller to init the master
+	err := hacluster.cluster.StartClusterMaster(hacluster.nodeName)
+	if err != nil {
+		return err
+	}
+
+	return hacluster.ovnController.Run()
+}
+
+//updateOvnDbEndpoints Updates the ovnkube-db endpoints. Should be called
+// only if ovnkube-master is leader. This function will create the ovnkube-db endpoints
+// if it doesn't exist.
+func (hacluster *OvnHAMasterController) updateOvnDbEndpoints(ep *kapi.Endpoints, configureInvalidIP bool) error {
+
+	var epIP string
+	if configureInvalidIP {
+		epIP = invalidIPAddress
+	} else {
+		epIP = config.Kubernetes.PodIP
+	}
+
+	epSubsets := []v1.EndpointSubset{
+		{
+			Addresses: []v1.EndpointAddress{
+				{IP: epIP},
+			},
+			Ports: []v1.EndpointPort{
+				{
+					Name: "north",
+					Port: ovnNorthDbPort,
+				},
+				{
+					Name: "south",
+					Port: ovnSouthDbPort,
+				},
+			},
+		},
+	}
+
+	var err error
+	if ep == nil {
+		logrus.Debugf("updateOvnDbEndpoints : Creating the endpoint")
+		// Create the endpoint
+		ovndbEp := v1.Endpoints{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: config.Kubernetes.OVNConfigNamespace,
+				Name:      ovnkubeDbEp,
```

Comment:
@numansiddique instead of SetAnnotations() below, you should be able to:

```
			ObjectMeta: metav1.ObjectMeta{
				Namespace: config.Kubernetes.OVNConfigNamespace,
				Name:      ovnkubeDbEp,
				Annotations: map[string]string{
					"ovnkube-master-leader": hacluster.NodeName,
				},
```

Reply:
Done